Register QEMU with binfmt_misc filesystem

smacfarlane opened the pull request.

samsalisbury commented on scripts/assertions.bash:
```diff
@@ -51,3 +51,15 @@ assert_tarball_not_contains_tags() { TARBALL="$1"; shift
+assert_is_binfmt_registered() { 
+  [ -f "/proc/sys/fs/binfmt_misc/$1" ] && return 0
+  return 1
+}
```
I like the explicitness of this, but I think BATS automatically has `set -e` enabled so the returns may not be necessary? Happy either way though, and I could be wrong, Bash is anything but uniform and predictable in behaviour sometimes.

samsalisbury commented on scripts/register_qemu_binfmt.bats:
```diff
@@ -0,0 +1,44 @@
+  #  CI.  If it is desireable to run this particular test locally, it is recommended to
+  #  obtain a VM (ex: with Vagrant) and run this test suite with `CI=true` set in the 
+  #  environment
+  skip_unless_ci
```
I am 100% behind gating this test due to its destructive potential. And thanks for the comment!

In fact, maybe even in CI we don't want this necessarily running alongside other tests, as it seems it could affect them? (Especially if running them in parallel which is a nice option to keep hold of for later optimisation.) Could we gate with an explicit opt-in flag maybe e.g. `RUN_DESTRUCTIVE_TESTS=true`?

samsalisbury commented on scripts/register_qemu_binfmt.bats:
```diff
@@ -0,0 +1,44 @@
+  assert_binfmt_fix_binary_flag_is_set qemu-s390x
+  echo "The 'F' flag is set for qemu-arm"
+  assert_binfmt_fix_binary_flag_is_set qemu-arm
+}  
```
Even though these are all just assertions on the same underlying state, I wonder if splitting them into separate tests would be helpful? By default BATS only prints the test names if they pass, so all these echoes will mostly not be seen, but as test names they would be.

smacfarlane commented on action.yml:
```diff
@@ -19,6 +19,11 @@ inputs:
       Each tag must be in the format: <registry>/[<namespace>/]<repository>:<tag>
     required: true
 
+  artifactory_token:
```
We don't have access to the `secrets` context in a composite action.  Having to pass it in as a required field doesn't feel great, but I can't think of another option.  Is there a better way to handle this?

samsalisbury replied:
Hm, I didn't realise this change would require authentication. Because this is a public action, it needs to be able to run on GitHub-hosted runners, which don't have access to Vault. Therefore it would mean we'd need to add a GitHub secret to the org, or each project needing it.

Is there a reason we want to keep the Docker image private? Would it not be possible to push it to a public repository so we can avoid needing authentication?

samsalisbury replied:
(If authentication ends up being unavoidable, passing it in this way would be fine IMO, there aren't really any other choices when using GitHub-hosted runners.)

smacfarlane replied:
I've got a request in with prodsec to see if we can move where we're hosting that image.

smacfarlane commented on action.yml:
```diff
@@ -96,6 +102,17 @@ runs:
 
       run: ${{ github.action_path }}/scripts/digest_inputs 
 
+    - name: Docker Login - docker.artifactory.hashicorp.engineering(RO)
```
This is an end-of-week 🧠 question/thought,  will logging in to the Artifactory docker mirror here have any side-effects in the calling workflows?

samsalisbury replied:
For steps in the same job, the login will persist, so yes it would cause the calling job to remain logged in. That by itself is probably not too much of a problem. However, in general having this action reliant on authentication is something we want to avoid I think (see above).

smacfarlane commented:
@samsalisbury I think I addressed all your feedback (thank you!).  I've left a couple in-line questions that may be non-issues but wanted to make sure I left my notes before leaving for the weekend.

samsalisbury commented on scripts/assertions.bash:
```diff
@@ -51,3 +51,13 @@ assert_tarball_not_contains_tags() { TARBALL="$1"; shift
+
+assert_binfmt_fix_binary_flag_is_set() {
+  local flags
+  IFS=":" read _ flags <<< $(grep "flags" /proc/sys/fs/binfmt_misc/$1)
```
What does the underscore do in `read _ ...` ? Would there be a case for adding the `-r` (raw) flag here? (Shellcheck has me trained up to always use that flag, but I'm sure there are cases when it's not appropriate.)

smacfarlane replied:
I use that as `not used`, out of habit from other languages, though this question make me wonder if it had special meaning in bash. Which it does but it still probably safe since we don't use the value.  I'll update it to be a `local notused` though to make it more explicit and safe. 

I think it's the same either way with `-r` for this case. The line we're looking for is `flags: <FLAGS>` where `<FLAGS>` is some combination of `P`, `G`, `C`, and `F`, and there is only one `flags: ` line in that file.

samsalisbury replied:
Ah, thanks! So we're reading two fields, not interested in the first, and expecting the second to be the flags field... I haven't seen read used in this way before, but makes sense now. Have you considered using the `cut` command to avoid needing to set IFS, something like this: ?

```bash
  flags="$(cut -d: -f2 <<< $(grep "flags" /proc/sys/fs/binfmt_misc/$1))"
```

I'm not sure if it's better or not, I like the explicitness of the `-f2` (field number 2) param and the `-d:` (delimiter colon) one, but now I understand read I think that's fine too.

smacfarlane replied:
🤔  I hadn't thought of doing it like that.  `read` is more clear to me, but that's mostly due to familiarity, and I do like the explicitness of the cut,  changed 😄

samsalisbury reviewed (commented):
Hi @smacfarlane, this is looking great, the only issue is needing to authenticate to Artifactory, which I hope we can find a way around. If it's possible to make the image public, maybe on ECR so there's no rate limit, then we'd be in a great place. If that's not possible, could we make the Dockerfile it's built from public, then we could build the image inline maybe? Let's talk about this later! :)
